cmd/remote-session.go

```diff
@@ -0,0 +1,222 @@
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	return cmd.Run()
+}
+
+// Destroys the "remote session". In reality it just deletes
+// the container referenced by $COREOS_ASSEMBLER_REMOTE_SESSION.
+func runDestroy(c *cobra.Command, args []string) error {
+	if err := checkEnv(); err != nil {
+		return err
+	}
+	session := os.Getenv("COREOS_ASSEMBLER_REMOTE_SESSION")
+	podmanargs := []string{"--remote", "--rm", "-f", session}
+	cmd := exec.Command("podman", podmanargs...)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	return cmd.Run()
+}
+
+// Executes a command in the "remote session". Mostly just a
+// `podman --remote exec`.
+func runExec(c *cobra.Command, args []string) error {
+	if err := checkEnv(); err != nil {
+		return err
+	}
+	session := os.Getenv("COREOS_ASSEMBLER_REMOTE_SESSION")
+	podmanargs := []string{"--remote", "exec", "-it", session, "cosa"}
+	podmanargs = append(podmanargs, args...)
+	cmd := exec.Command("podman", podmanargs...)
+	cmd.Stdin = os.Stdin
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	if err := cmd.Run(); err != nil {
+		if exitError, ok := err.(*exec.ExitError); ok {
+			// If the command failed let's exit with the same exitcode
+			// as the remotely executed process.
+			os.Exit(exitError.ExitCode())
+		} else {
+			return err
+		}
+	}
+	return nil
+}
+
+// Executes a `podman --remote ps -a --filter id=<container>`
+// to show the status of the remote running cosa container.
+func runPS(c *cobra.Command, args []string) error {
+	if err := checkEnv(); err != nil {
+		return err
+	}
+	session := os.Getenv("COREOS_ASSEMBLER_REMOTE_SESSION")
+	podmanargs := []string{"--remote", "ps", "-a",
+		fmt.Sprintf("--filter=id=%s", session)}
+	cmd := exec.Command("podman", podmanargs...)
+	cmd.Stdin = os.Stdin
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	return cmd.Run()
+}
+
+// runSync provides an rsync-like interface that allows
+// files to be copied to/from the remote. It uses
+// `podman --remote exec` as the transport for rsync (see [1])
+//
+// One of the arguments here must be prepended with a `:`. This
+// argument will represent the path on the remote. This function
+// will substitute `:` with `$COREOS_ASSEMBLER_REMOTE_SESSION:`.
+// That environment var just contains the container ID on the remote.
+//
+// [1] https://github.com/moby/moby/issues/13660
+func runSync(c *cobra.Command, args []string) error {
+	// check environment
+	if err := checkEnv(); err != nil {
+		return err
+	}
+	// check arguments. Need one with pre-pended ':'
+	found := 0
+	for index, arg := range args {
+		if strings.HasPrefix(arg, ":") {
+			args[index] = fmt.Sprintf("%s%s",
+				os.Getenv("COREOS_ASSEMBLER_REMOTE_SESSION"), arg)
+			found++
+		}
+	}
+	if found != 1 {
+		return fmt.Errorf("Must pass in a single arg to rsync with `:` prepended")
+	}
+	// build command and execute
+	rsyncargs := []string{"-avh", "--mkpath", "--blocking-io",
+		"--rsh", "podman --remote exec --user root -i"}
```

Comment:
Pretty cool that this works!

Reply:
yup. This is some magic that is awesome. `podman` and `docker` are such awesome tools!